VDA-StorageNode-Registry/test/index.test.ts

```diff
@@ -0,0 +1,939 @@
+                        contract.removeDatacenter(invalidIds[i])
+                    ).to.be.revertedWithCustomError(contract, "InvalidDatacenterId").withArgs(invalidIds[i]);
+                }
+            })
+
+            it("Success: Fresh datacenters that has no storage nodes added", async () => {
+                const tx = await contract.removeDatacenter(datacenterIds[0]);
+
+                await expect(tx).to.emit(contract, "RemoveDataCenter").withArgs(datacenterIds[0]);
+            })
+
+            it("Failed: Removed datacenterId", async () => {
+                await expect(
+                    contract.removeDatacenter(datacenterIds[0])
+                ).to.be.revertedWithCustomError(contract, "InvalidDatacenterId");
+            })
+
+            it("Failed: Has depending nodes", async () => {
+                storageNode.datacenterId = datacenterIds[1];
+                // Add storage node 
+                await checkAddNode(storageNode, user, trustedSigner, true);
+
+                // Failed to remove datacenter
+                await expect(contract.removeDatacenter(datacenterIds[1])).to.be.revertedWithCustomError(contract, "HasDependingNodes");
+            })
+
+            it("Success: After depending nodes are removed",async () => {
+                // Remove start
+                const blockTime = await time.latest();
+                const unregisterTime = blockTime + days(30);
+                await checkRemoveNodeStart(user, unregisterTime);
+                
+                // Remove complete
+                await time.increaseTo(unregisterTime);
+                await checkRemoveNodeComplete(user);
+
+                // Success to remove datacenter
+                await expect(
+                    contract.removeDatacenter(datacenterIds[1])
+                ).to.emit(contract, "RemoveDataCenter").withArgs(datacenterIds[1]);
+            })
+        })
+    });
+
+    describe("Storage node", () => {
+
+        let snapShotWithDatacenters: SnapshotRestorer
+
+        const trustedSigner = Wallet.createRandom();
+        
+        const datacenterIds : BigNumber[] = [];
+        let maxDatacentID : BigNumber;
+
+        before(async () => {
+            await snapShotAfterDeploy.restore();
+
+            // Add datacenters
+            for (let i = 0; i < dataCenters.length; i++) {
+                const tx = await contract.addDatacenter(dataCenters[i])
+
+                const transactionReceipt = await tx.wait();
+                const event = transactionReceipt.events?.find(item => {
+                    return item.event === 'AddDataCenter'
+                })
+                if (event !== undefined) {
+                    datacenterIds.push(event.args![0])
+                }
+            }
+
+            maxDatacentID = datacenterIds[datacenterIds.length -1];
+
+            snapShotWithDatacenters = await takeSnapshot();
+        })
+
+        describe("Add storage node", () => {
```

Comment:
Add check for invalid didAddress.

Add check that `establishmentDate` is set correctly.

Add check that `getNodeByAddress` and `getNodeByEndpoint` return an extra `status` attribute (VIP-10 was updated).

Reply:
- Added check for invalid didAddresses. 
- Not able to check the correctness of `establishmentDate`. In the solidity, it set the `block.timestamp` as the `establishmentDate`. This `block.timestamp` is not representing the current time.
- `status` value check added in `getNodeByAddress` & `getNodeByEndpoint()` test